### ISSUE TITLE
FR Use funsor distributions directly in Pyro models

### ISSUE THREAD

**Issue by fritzo** [pyro-ppl/pyro#2618]
This issue proposes a low-level refactoring of `pyro.distribution` so that the basic object is no longer a `TorchDistribution` but a `Funsor` (or possibly a derived `class FunsorDistribution(Funsor, TorchDistribution)`). This low-level refactoring could precede #2580 insofar as it would be compatible with existing Pyro inference algorithms, and should be forwards compatible with #2580.

The motivation is to allow small funsor computations inside Pyro models using the `pyro.factor()` statement.

## Motivating Example 1: Beta-Binomial conjugacy

It would be nice to perform small Funsor computations directly in a Pyro model, using string notation:
```diff
  def model(data):
-     probs = pyro.sample("probs", dist.Beta(2, 2))
-     pyro.sample("obs", dist.Binomial(10, probs), obs=data)
+     prior = dist.Beta(2, 2).log_prob("probs")
+     likelihood = dist.Binomial(10, "probs").log_prob(data)
+     pyro.factor("obs", (prior + likelihood).reduce(ops.logaddexp))
```
or even simply
```diff
+     pyro.factor("obs", prior + likelihood)
```
where `pyro.factor` automatically calls `.reduce(ops.logaddexp)`, or could call `to_data()` which would in turn call `sum_product` which would `.reduce()` with either `ops.add` for plates or `ops.logaddexp` for non-plates.

## Motivating Example 2: Likelihoods vs Distributions

In some cases there is opportunity for significant caching of computation in a likelihood, e.g.
```py
class Model(PyroModule):
    def __init__(self, ...):
        data = torch.tensor(...)
        eager_param = torch.tensor(...)
        self.likelihood = lambda lazy_param: MyDistribution(eager_param, lazy_param).log_prob(data)
    def forward(self, ...):
        pyro.factor("likelihood", self.likelihood(lazy_param=torch.tensor(...))
```
A real world example is Pyro's [CoalescentTimesWithRate](http://docs.pyro.ai/en/stable/distributions.html#pyro.distributions.CoalescentTimesWithRate) distribution and the [CoalescentRateLikelihood](http://docs.pyro.ai/en/dev/contrib.epidemiology.html#pyro.distributions.CoalescentRateLikelihood). These two classes would most naturally be unified as a single `Funsor` class with different `eager` reduction rules depending on which arguments were lazy: (`coal_times` versus `rate_grid`)

## What would be required?
@eb8680 advice appreciated :smile:
- [ ] Make funsor a top-level dependency of Pyro (requires a funsor release?, stabilized api?)
- [ ] Ensure `FunsorDistribution.log_prob()` returns a ground `torch.Tensor` rather than a `funsor.Tensor` in case there are no free variables
- [ ] Support `.to_event()`
- [ ] Support `.mask()` (I think this requires a `MaskFunsor`?)
- [ ] Wrap existing distributions in `FunsorDistribution`
- [ ] Move some Pyro distributions into Funsor?

### First PR:

<details>
 
```python
def to_funsor_dist(cls):
    return getattr(funsor.torch.distributions,
                   cls.__name__.split("Wrapper_")[-1])

class FunsorizeMeta(ABCMeta):  # because
    def __init__(cls, name, supers, attrs):
        # Wrap .log_prob() method to return a Funsor.
        if "log_prob" in attrs:
            base_log_prob = attrs["log_prob"]
        def log_prob(self, value):
            if isinstance(value, Funsor):
                return to_funsor(self)(value=value)
            return base_log_prob(self, value)
        attrs["log_prob"] = log_prob
        return super().__init__(cls, name, supers, attrs)
    def __call__(cls, *args, **kwargs):
        # TODO support string inputs?
        if any(isinstance(f, Funsor) for f in ar for a in (args, kwargs.values())):
            return to_funsor_dist(cls)(*args, **kwargs)
        return super().__call__(*args, **kwargs)
```
Then we can make the following one-line diff
```diff
- class TorchDistributionMixin(Distribution):
+ class TorchDistributionMixin(Distribution, metaclass=FunsorizeMeta):
```
then add a beta-binomial unit test:
```python
def model(data):
    probs = funsor.Variable("probs", funsor.Real)
    prior = dist.Beta(2, 2).log_prob(probs)
    likelihood = dist.Binomial(10, probs).log_prob(data)
    pyro.factor("obs", (prior + likelihood).reduce(ops.logaddexp))
```

</details>

### Second PR:

<details>

```python
class LazySectionMessenger(TraceMessenger):

    def _process_message(self, msg):
        super()._process_message(msg)
        if msg["type"] == "sample":
            msg["stop"] = True
            
    def _pyro_sample(self, msg):
        if site["value"] is not None:
            site["value"] = funsor.Variable(msg["name"], msg["fn"].value_domain)
        site["done"] = True
        
    def __enter__(self, ...):
        self.preserved_plates = frozenset(h.name for h in _PYRO_STACK
                                     if isinstance(h, pyro.plate))
        return super().__enter__()
            
    def __exit__(self, *args):
        # Convert delayed statements to pyro.factor()
        super().__exit__(*args)
        log_prob_terms, reduced_vars = [], []
        for name, site in self.trace.items():
            if not site["is_observed"]:
                reduced_vars.append(name)
            log_prob = site["fn"].log_prob(site["value"])  # Requires #2618
            plates |= frozenset(f.name for f in site["cond_indep_stack"]
                                if f.vectorized)
        if log_prob_terms:
            reduced_plates = plates - self.preserved_plates
            log_prob = funsor.sum_product.sum_product(
                funsor.ops.logaddexp,
                funsor.ops.add,
                log_prob_terms,
                eliminate=frozenset(reduced_vars) | reduced_plates,
                plates=plates,
            )
            name = reduced_vars[0]
            pyro.factor(name, log_prob)
```
Then add a unit test
```python
def model_1(data):
    with lazy_section():
        probs = pyro.sample("probs", dist.Beta(2, 2))
        with pyro.plate("plate", len(data)):
            # FIXME validation will fail due to .batch_shape.
            # FIXME the plate's call of .expand() will break.
            pyro.sample("obs", dist.Binomial(5, probs),
                        obs=data)
def model_2(data):
    with lazy_section():
        with pyro.plate("plate", len(data)):
            # FIXME the plate's call of .expand() will break.
            probs = pyro.sample("probs", dist.Beta(2, 2))
            # FIXME validation will fail due to .batch_shape.
            # FIXME the plate's call of .expand() will break.
            pyro.sample("obs", dist.Binomial(5, probs),
                        obs=data)
```

</details>

**Comment by eb8680** [comment]
One design question is the extent to which this should involve new Pyro/Funsor hybrid syntax versus just using the Funsor distribution wrappers.  I think a version of example 1 should already work, and we don't need to add anything new to `FunsorDistribution`. 
```py
prior = funsor.distributions.Beta(2, 2)(value="probs")
likelihood = funsor.distributions.Binomial("probs")(value="data")
pyro.factor("obs", to_data((prior + likelihood)(data=to_funsor(data))))
```
We could also add a `ground_funsors` handler that automates the conversion call in the last line and ensures that all Funsor distributions and values are converted to data at `sample` and `factor` sites.

**Comment by fritzo** [comment]
Closing this as mostly resolved by `poutine.collapse`. Further minor improvements can be deferred to follow-up issues.
